### PR TITLE
Protocol2.2

### PR DESCRIPTION
**Pull request by LordAro**
Bring playd up to spec with current protocolv2

### CODE REVIEW

**Comment by LordAro** [comment]
oh, damn, i never even tried to fix the tests

**Comment by CaptainHayashi** [comment]
@LordAro **fix**

File: src/audio/audio.cpp [hunk 1]
```diff
@@ -40,15 +37,14 @@ Audio::State NoAudio::Update()
...
-std::unique_ptr<Response> NoAudio::Emit(const std::string &path, bool)
+std::pair<std::string, std::string> NoAudio::Emit(const std::string &path)
 {
-	std::unique_ptr<Response> ret;
-
-	if (path == "/control/state") {
-		ret = Response::Res("Entry", "/control/state", "Ejected");
+	if (path == "/player/state/current" || path == "/player/state/available") {
+		// Only possible value is the set value
+		return std::make_pair("Entry", "ejected");
```
Comment (CaptainHayashi) [inline src/audio/audio.cpp]: Seems strange that this isn't `AUDIO::STATE_STRINGS[Audio::State::NONE]`, in terms of DRY

File: src/audio/audio.cpp [hunk 2]
```diff
@@ -87,23 +87,23 @@ std::unique_ptr<Response> PipeAudio::Emit(const std::string &path,
 
 	std::string value;
 
-	if (path == "/control/state") {
```
Comment (CaptainHayashi) [inline src/audio/audio.cpp]: `/control/state` will need to be a thing for compliance, but it'll just return `running` every time

File: src/audio/audio.cpp [hunk 3]
```diff
@@ -87,23 +87,23 @@ std::unique_ptr<Response> PipeAudio::Emit(const std::string &path,
...
+	if (path == "/player/state/current") {
 		auto state = this->sink->State();
-		auto playing = state == Audio::State::PLAYING;
-		value = playing ? "Playing" : "Stopped";
+		value = Audio::STATE_STRINGS[static_cast<int>(state)];
+	} else if (path == "/player/state/available") {
+		// TODO: Use STATE_STRINGS?
```
Comment (CaptainHayashi) [inline src/audio/audio.cpp]: Probably

File: src/audio/audio.cpp [hunk 4]
```diff
@@ -87,23 +87,23 @@ std::unique_ptr<Response> PipeAudio::Emit(const std::string &path,
...
-		if (!can) return ret;
 		value = std::to_string(micros);
-	} else return ret;
+	} else if (path == "/player/time/total") {
+		std::uint64_t micros = this->TotalLength();
+		value = std::to_string(micros);
+	}
 
-	return Response::Res("Entry", path, value);
+	return std::make_pair("Entry", value);
```
Comment (CaptainHayashi) [inline src/audio/audio.cpp]: First value of the pair should be `integer` for `/player/time/total` and `/player/time/elapsed`, and `string` for everything else, see [the types RFC](https://github.com/UniversityRadioYork/bifrost-rfcs/blob/master/protocol/types.md) and [the draft player RFC](https://github.com/UniversityRadioYork/bifrost-rfcs/blob/mw-player/resource-hierarchy/player.md).

File: src/audio/audio.hpp [hunk 5]
```diff
@@ -105,6 +105,19 @@ class Audio
 	 * @see Seek
 	 */
 	virtual std::uint64_t Position() const = 0;
+
+	/**
+	 * This Audio's length.
+	 * @return Audio's length, in microseconds.
+	 */
+	virtual std::uint64_t TotalLength() const = 0;
```
Comment (CaptainHayashi) [inline src/audio/audio.hpp]: Something for later might be figuring out if this should be an `int64_t`, given how other things may be using signed integers only.

File: src/audio/sources/mp3.hpp [hunk 6]
```diff
@@ -51,16 +51,20 @@ class Mp3AudioSource : public AudioSource
...
-	/// The size of the internal decoding buffer, in bytes.
-	static const size_t BUFFER_SIZE;
+	/** The size of the internal decoding buffer, in bytes.
+	 * This value is somewhat arbitrary, but corresponds to the minimum buffer size
+	 * used by ffmpeg, so it's probably sensible.
+	 */
+	static const size_t BUFFER_SIZE = 16384;
 
-	std::vector<std::uint8_t> buffer; ///< The decoding buffer.
+	std::array<std::uint8_t, BUFFER_SIZE> buffer; ///< The decoding buffer.
```
Comment (CaptainHayashi) [inline src/audio/sources/mp3.hpp]: Nice catch, btw

File: src/io.cpp [hunk 7]
```diff
@@ -237,19 +237,7 @@ void IoCore::Shutdown()
...
 
 	// Finally, kill off all of the connections with 'fatal' responses.
-	for (const auto conn : this->pool) IoCore::TryShutdown(conn);
-}
-
-/* static */ void IoCore::TryShutdown(const std::shared_ptr<Connection> conn)
-{
-	if (!conn) return;
-
-	auto response = Response(Response::Code::STATE).AddArg("Quitting");
```
Comment (CaptainHayashi) [inline src/io.cpp]: I presume this is because nothing is ever likely to read this before libuv shuts down their connections?

File: src/io.hpp [hunk 8]
```diff
@@ -90,10 +90,10 @@ class IoCore : public ResponseSink
 
 	void Respond(const Response &response, size_t id = 0) const override;
 
-private:
 	/// The period between player updates.
 	static const uint16_t PLAYER_UPDATE_PERIOD;
```
Comment (CaptainHayashi) [inline src/io.hpp]: Does this still need to be public?

File: src/messages.h [hunk 9]
```diff
@@ -15,12 +15,15 @@
...
 
-/// Message shown when a client connects to playd.
+/// Playd's version string. Used in OHAI
 #ifndef PD_VERSION
 #define PD_VERSION "0.0.0"
 #endif
 const std::string MSG_OHAI = "playd " PD_VERSION;
 
+// Protocol version supported by Playd. Used in OHAI.
+const std::string MSG_PROTO_VER = "bifrost α";
```
Comment (CaptainHayashi) [inline src/messages.h]: ou

File: src/audio/audio.hpp (hunk 5 above)
Comment (LordAro) [inline src/audio/audio.hpp]: I figured since Position() used a uint64_t, it would make sense to use the same for this

Comment (CaptainHayashi) [inline src/audio/audio.hpp]: Ah, I meant those too.  Another problem for another day though.

**Comment by LordAro** [comment]
superseded by #108

**Pull request closed without merging**